src/linker/configuration.py

```diff
@@ -140,10 +154,30 @@ def _load_input_data_paths(input_data: Path) -> List[Path]:
         return file_list
 
     @staticmethod
-    def _get_spark_requests(
-        environment: Dict[str, Union[Dict, str]]
-    ) -> Optional[Dict[str, Any]]:
-        spark = environment.get("spark", None)
-        if spark and not "keep_alive" in spark:
-            spark["keep_alive"] = False
-        return spark
+    def _assign_default_environment(
+        subenv: Dict[str, Any], defaults: Dict[str, Any]
+    ) -> Dict[str, Any]:
+        """Recursively fill in missing values with defaults."""
+        required_keys = ["computing_environment", "container_engine"]
+        outer_keys = DEFAULT_ENVIRONMENT_VALUES.keys()
+        for key, default_value in defaults.items():
+            if key in required_keys and key not in subenv:
+                # Add missing required items
+                logger.info(f"Assigning default value '{default_value}' to '{key}'")
+                subenv[key] = default_value
+            if key not in required_keys and key in subenv:
+                # Update if outer key is present
+                if isinstance(default_value, dict) and isinstance(subenv[key], dict):
```

Comment:
You don't need to worry about one of them being a dict and the other not?

Reply:
Hmm, good catch. What that means at that point is that the user is simply doing something incorrect, i.e. providing a dict when there shouldn't be or a non-dict when there should be. I'll raise for now and then this should be properly validated in a future ticket.